File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        def build(in_start, in_end, post_index):
            if in_start > in_end:
                return None

            val = postorder[post_index[0]]
            post_index[0] -= 1
            root = TreeNode(val)

            in_index = inorder_map[val]
            root.right = build(in_index + 1, in_end, post_index)
            root.left = build(in_start, in_index - 1, post_index)

            return root

        inorder_map = {val: idx for idx, val in enumerate(inorder)}
        return build(0, len(inorder) - 1, [len(postorder) - 1])

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        int n = inorder.length;
        return build(inorder, postorder, 0, n - 1, new int[]{n - 1});
    }

    private TreeNode build(int[] inorder, int[] postorder, int inStart, int inEnd, int[] postIndex) {
        if (inStart > inEnd) return null;

        int val = postorder[postIndex[0]--];
        TreeNode root = new TreeNode(val);

        int inIndex = inStart;
        while (inorder[inIndex] != val) inIndex++;

        root.right = build(inorder, postorder, inIndex + 1, inEnd, postIndex);
        root.left = build(inorder, postorder, inStart, inIndex - 1, postIndex);

        return root;
    }
}